src/cata_ui.cpp

```diff
@@ -0,0 +1,442 @@
+{
+    auto o_rect = other.get_rect();
+    rect.x = o_rect.x - o_rect.size_x + x;
+    rect.y = o_rect.y + y;
+    set_anchor( other.get_anchor() );
+}
+
+void ui_element::set_rect( const ui_rect &new_rect )
+{
+    rect = new_rect;
+    calc_anchored_values();
+}
+
+ui_anchor ui_element::get_anchor() const
+{
+    return anchor;
+}
+
+void ui_element::set_anchor( ui_anchor new_anchor )
+{
+    anchor = new_anchor;
+    calc_anchored_values();
+}
+
+void ui_element::calc_anchored_values()
+{
+    if( parent != nullptr ) {
+        auto p_rect = parent->get_rect();
+        switch( anchor ) {
+            case ui_anchor::top_left:
+                break;
+            case ui_anchor::top_center:
+                anchored_x = ( p_rect.size_x / 2 ) - ( rect.size_x / 2 ) + rect.x;
+                anchored_y = rect.y;
+                return;
+            case ui_anchor::top_right:
+                anchored_x = p_rect.size_x - rect.size_x + rect.x;
+                anchored_y = rect.y;
+                return;
+            case ui_anchor::center_left:
+                anchored_x = rect.x;
+                anchored_y = ( p_rect.size_y / 2 ) - ( rect.size_y / 2 ) + rect.y;
+                return;
+            case ui_anchor::center_center:
+                anchored_x = ( p_rect.size_x / 2 ) - ( rect.size_x / 2 ) + rect.x;
+                anchored_y = ( p_rect.size_y / 2 ) - ( rect.size_y / 2 ) + rect.y;
+                return;
+            case ui_anchor::center_right:
+                anchored_x = p_rect.size_x - rect.size_x + rect.x;
+                anchored_y = ( p_rect.size_y / 2 ) - ( rect.size_y / 2 ) + rect.y;
+                return;
+            case ui_anchor::bottom_left:
+                anchored_x = rect.x;
+                anchored_y = p_rect.size_y - rect.size_y + rect.y;
+                return;
+            case ui_anchor::bottom_center:
+                anchored_x = ( p_rect.size_x / 2 ) - ( rect.size_x / 2 ) + rect.x;
+                anchored_y = p_rect.size_y - rect.size_y + rect.y;
+                return;
+            case ui_anchor::bottom_right:
+                anchored_x = p_rect.size_x - rect.size_x + rect.x;
+                anchored_y = p_rect.size_y - rect.size_y + rect.y;
+                return;
+        }
+    }
+    anchored_x = rect.x;
+    anchored_y = rect.y;
+}
+
+unsigned int ui_element::get_ax() const
+{
+    return anchored_x;
+}
+
+unsigned int ui_element::get_ay() const
+{
+    return anchored_y;
+}
+
+void ui_element::set_parent( const ui_window *parent )
+{
+    this->parent = parent;
+    calc_anchored_values();
+}
+
+const ui_rect &ui_element::get_rect() const
+{
+    return rect;
+}
+
+WINDOW *ui_element::get_win() const
+{
+    if( parent != nullptr ) {
+        return parent->get_win();
+    }
+    return nullptr;
+}
+
+ui_label::ui_label( std::string text ,int x, int y, ui_anchor anchor ) : ui_element( utf8_width( text.c_str() ), 1, x, y, anchor ), text( text )
```

Comment:
`utf8_width` now has an overload that takes a `std::string`, so you don't need `.c_str()` when calling it.